nrf_rpc/doc/architecture.rst

```diff
@@ -0,0 +1,135 @@
+.. architecture:
+
+
+Architecture
+############
+
+The following picture gives an overview of the nRF RPC architecture:
+
+.. image:: img/layers.svg
+   :alt: nRF RPC Architecture Layers
+   :align: center
+
+
+Types of communication
+======================
+
+Calls can be done in one of two ways: using a command/response or an event.
+
+**Command and response** are intended to be used for a synchronous function calls.
+The caller sends the command to the other side and waits for a response.
+The response is sent after the remote function returns on the remote core.
+The following image shows an example of the command and response flow.
+
+.. image:: img/cmd_flow.svg
+   :alt: nRF RPC Command/Response flow
+   :align: center
+
+**Events** are intended to be used for an asynchronous function calls.
+The caller sends an event to the other side and returns immediately if there is free thread in thread pool (see :ref:`nrf_rpc_architecture_threads` section), otherwise it waits for available thread.
+It is not possible to return anything from the remote side after receiving an Event, but it is possible to send an Event in opposite direction.
+The following image shows an example of the event flow.
+
+.. image:: img/evt_flow.svg
+   :alt: nRF RPC Event flow
+   :align: center
+
+.. _nrf_rpc_architecture_threads:
+
+Threads
+=======
+
+When a remote procedure call is received, it must be executed in context of a thread.
+For this reason each side of the nRF RPC contains a thread pool.
+It is OS-dependent and it is implemented in OS abstraction layer.
+Number of threads in the thread pool is configurable.
+
+When a caller wants to call a remote procedure, the nRF RPC checks if there is a thread available in a pool.
+If there is not, then it waits until one becomes available, blocking the caller.
+After that nRF RPC sends a packet and it is executed by available thread from a thread pool.
+In the case of a Command being received, a Response is send directly to a waiting thread, so no new thread is allocated for a response.
+
+Sample command/response diagrams are on the image below.
+
+.. image:: img/cmd_simple.svg
+   :alt: nRF RPC Simple Command flow
+   :align: center
+
+A thread waiting for a response may be reused to receive a new incoming command from the remote thread that the local thread is waiting for, e.g. when callback is called synchronously.
+Below diagram shows that situation.
+
+.. image:: img/cmd_recursive.svg
+   :alt: nRF RPC Simple Command flow
+   :align: center
+
+Events always reserve a new thread from remote thread pool.
+Special attention is required when sending multiple events one after another, because each event will reserve a new thread. Sending events too fast may consume entire thread pool and as the result block all the outgoing commands and events.
+Sample events are shown on diagram below.
+
+.. image:: img/evt_simple.svg
+   :alt: nRF RPC Simple Command flow
+   :align: center
+
+
+Error handling
```

Comment:
It's not totally clear to me here when error messages are sent between cores, and when they are just sent as callbacks to the local user.

Reply:
They are always send as a callback to the local user. Additionally they are send to the remote user if the error occurred during receiving/parsing the incoming packet.